From 35077904b779b558faa883d7ba6c664fe5eb4906 Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Tue, 10 Aug 2021 14:46:29 +0300
Subject: [PATCH] Merge pull request #3447 in SW_UGW/linux from
 feature/DRV_DPM_SW-520-add-bridging-domains-for-the-different-pon-traffic-types-uc-mc-bc
 to master

* commit 'd8cdda08f560ee870fd83b31731ff236a7c63215':
  DRV_DPM_SW-520 - add bridging domains for different traffic types
---
 include/net/datapath_api.h  | 27 +++++++++++++++++++++++++++
 include/net/datapath_inst.h |  1 +
 2 files changed, 28 insertions(+)

diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index c7a966c41a9e..ec32c8c66fa9 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -869,6 +869,9 @@ enum DP_SUBIF_DATA_FLAG {
 					 *   set this device's mac address to
 					 *   CPU bridge port
 					 */
+	DP_SUBIF_BR_DOMAIN = BIT(14), /*!< once this flag is set, it means
+				       * domain_id and domain_members are valid
+				       */
 };
 
 /*! @brief dp_subif_id struct for get_netif_subif */
@@ -961,6 +964,16 @@ struct dp_subif_data {
 	u16 swdev_en_flag; /*!< [in] swdev_en_flag =1 - Support h/w offload
 			    * swdev_en_flag = 0 - no h/w offload support
 			    */
+	u8 domain_id:5;	/*!< [in] we support 32 domain per bridge only
+			 * valid only DP_SUBIF_BR_DOMAIN is set in flag_ops
+			 */
+	u32 domain_members; /*!< [in] one bit for one domain_id:
+			     * bit 0 for domain_id 0, bit 1 for domain 1 and so on.
+			     * If one bit is set to 1, the traffic received
+			     * from this dev can be forwarded to this domain
+			     * as specified in domain_members
+			     * valid only if DP_SUBIF_BR_DOMAIN is set in flag_ops
+			     */
 };
 
 /*! @brief enum DP_F_DATA_RESV_CQM_PORT */
@@ -2699,6 +2712,20 @@ struct dp_subif_upd_info {
 			      * for GPON, always 1, for EPON, 1-8
 			      */
 	u32 flags; /*!< [in] caller provided during dp_update_subif_info */
+	u8 new_domain_flag : 1; /*!< [in] if set, then use domain
+				 * information, otherwise keep original
+				 * domain information
+				 */
+	u8 new_domain_id : 5; /*!< [in] we support 32 domain per bridge only.
+			       * valid only DP_SUBIF_BR_DOMAIN is set in flag_ops
+			       */
+	u32 domain_members; /*!< [in] one bit for one domain_id:
+			     * bit 0 for domain_id 0, bit 1 for domain 1,so on.
+			     * If one bit is set to 1, the traffic received
+			     * from this dev can be forwarded to this domain
+			     * as specified in domain_members
+			     * valid only if DP_SUBIF_BR_DOMAIN is set in flag_ops
+			     */
 };
 
 /*!
diff --git a/include/net/datapath_inst.h b/include/net/datapath_inst.h
index 026a16d4cae2..942d59f6e181 100644
--- a/include/net/datapath_inst.h
+++ b/include/net/datapath_inst.h
@@ -150,6 +150,7 @@ struct inst_info {
 	int (*dp_set_ctp_bp)(int inst, int ctp, int portid, int bp,
 			     struct subif_platform_data *data);
 	int (*dp_get_queue_qocc)(struct dp_qos_queue_info *info, int flag);
+	int (*dp_cfg_domain_for_bp_member)(int inst, int bp);
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_SWITCHDEV)
 	int swdev_flag;
 	/* VLAN AWARE feature flag currently enabled only for PRX */
